lib/Sema/ConstraintSystem.h

```diff
@@ -2228,51 +2228,77 @@ class ConstraintSystem {
     Error
   };
 
+  class TypeMatchResult {
+    SolutionKind Kind;
+
+  public:
+    inline bool isSuccess() const { return Kind == SolutionKind::Solved; }
+    inline bool isFailure() const { return Kind == SolutionKind::Error; }
+    inline bool isAmbiguous() const { return Kind == SolutionKind::Unsolved; }
+
+    static TypeMatchResult success(ConstraintSystem &cs) {
```

Comment:
You may want to consider making helpers on ConstraintSystem like:
```
  TypeMatchResult getTypeMatchSuccess() {
    return TypeMatchResult::success(*this);
  }
  TypeMatchResult getTypeMatchFailure(ConstraintLocatorBuilder *locator) {
    return TypeMatchResult::failure(*this, locator);
  }
```

That would make some the other changes simpler/cleaner.

Reply:
Good point! Working on that right now.